core/lib/spree/core/controller_helpers/order.rb

```diff
@@ -58,7 +58,22 @@ def associate_user
           session[:guest_token] = nil
         end
 
+        def remove_invalid_session_order_id
+          user = try_spree_current_user
+
+          if session[:order_id]
+            order = Spree::Order.find(session[:order_id])
+            if user && order.user != user
+              session[:order_id] = nil
+            elsif !user && && order.user_id
```

Comment:
I think @matiasleidemer is right and, if I may suggest a refactor I'd go with this:

```ruby
def remove_invalid_session_order_id
  if session[:order_id]
    user = try_spree_current_user
    order = Spree::Order.find(session[:order_id])

    if (user && order.user != user) || (!user && order.user_id)
      session[:order_id] = nil
    end
  end
end
```